pkg/policy/k8s/service.go

```diff
@@ -0,0 +1,321 @@
+	if svc == nil {
+		return false
+	}
+	return p.config.EnableHighScaleIPcache || svc.IsExternal()
+}
+
+// onServiceEvent processes a ServiceNotification and (if necessary)
+// recalculates all policies affected by this change.
+func (p *PolicyWatcher) onServiceEvent(event k8s.ServiceNotification) {
+	err := p.updateToServicesPolicies(event.ID, event.Service, event.OldService)
+	if err != nil {
+		p.log.WithError(err).WithFields(logrus.Fields{
+			logfields.Event:     event.Action,
+			logfields.ServiceID: event.ID,
+		}).Warning("Failed to recalculate CiliumNetworkPolicy rules after service event")
+	}
+}
+
+// updateToServicesPolicies is to be invoked when a service has changed (i.e. it was
+// added, removed, its endpoints have changed, or its labels have changed).
+// This function then checks if any of the known CNP/CCNPs are affected by this
+// change, and recomputes them by calling resolveCiliumNetworkPolicyRefs.
+func (p *PolicyWatcher) updateToServicesPolicies(svcID k8s.ServiceID, newSVC, oldSVC *k8s.Service) error {
+	var errs []error
+
+	// Bail out early if updated service is not selectable
+	if !(p.isSelectableService(newSVC) || p.isSelectableService(oldSVC)) {
+		return nil
+	}
+
+	// newService is true if this is the first time we observe this service
+	newService := oldSVC == nil
+	// changedService is true if the service label or selector has changed
+	changedService := !newSVC.DeepEqual(oldSVC)
+
+	// candidatePolicyKeys contains the set of policy names we need to process
+	// for this service update. By default, we consider all policies with
+	// a ToServices selector as candidates.
+	candidatePolicyKeys := p.toServicesPolicies
+	if !(newService || changedService) {
+		// If the service definition itself has not changed, and it's not the
+		// first time we process this service, we only need to check the
+		// policies which are known the select the old version of the service
+		candidatePolicyKeys = p.cnpByServiceID[svcID]
+	}
+
+	// Iterate over all policies potentially affected by this service update,
+	// and re-resolve the policy refs for each.
+	for key := range candidatePolicyKeys {
+		cnp, ok := p.cnpCache[key]
+		if !ok {
+			p.log.WithFields(logrus.Fields{
+				logfields.Key:       key,
+				logfields.ServiceID: svcID,
+			}).Error("BUG: Candidate policy for service update not found. Please report this bug to Cilium developers.")
+			continue
+		}
+
+		// Skip policies which are not affected by this service update
+		if !(p.cnpMatchesService(cnp, svcID, newSVC) ||
+			(!newService && changedService && p.cnpMatchesService(cnp, svcID, oldSVC))) {
+			continue
+		}
+
+		p.log.WithFields(logrus.Fields{
+			logfields.CiliumNetworkPolicyName: cnp.Name,
+			logfields.K8sAPIVersion:           cnp.APIVersion,
+			logfields.K8sNamespace:            cnp.Namespace,
+			logfields.ServiceID:               svcID,
+		}).Debug("Service updated or deleted, recalculating CiliumNetworkPolicy rules")
```

Comment:
Can we protect this debug call with a log level check?  If this function is called when Services are updated, then depending on the amount of churn the cluster faces this debug statement could become expensive. See https://github.com/cilium/cilium/pull/26327 for context. We're currently working on a way to fix this long-term without this kind of workaround, but for right now it's a good idea to add this kind of log-level check for debug statements in potential hot paths.